### PR TITLE
[MOB-2028] JWT bindings

### PR DESCRIPTION
**Pull request by davidtruong**
JWT auth support for React Native.

### CODE REVIEW

**Review by tapashmajumder — APPROVED** [review]
+1

This is a provisional +1. And only for iOS and React Native. I have not reviewed Android.

Jay please do not merge until you have tested JWT end to end.

File: android/build.gradle [hunk 1]
```diff
@@ -22,5 +28,5 @@ def getModuleVersion() {
 
 dependencies {
     implementation 'com.facebook.react:react-native:+'
-    api 'com.iterable:iterableapi:3.2.6'
+    api 'com.github.Iterable.iterable-android-sdk:iterableapi:3.2.7-beta2'
```
Comment (Ayyanchira) [inline android/build.gradle:31]: Need changes to accommodate the latest GA build

File: android/src/main/java/com/iterable/reactnative/RNIterableAPIModule.java [hunk 2]
```diff
@@ -422,9 +436,29 @@ public boolean handleIterableURL(@NonNull Uri uri, @NonNull IterableActionContex
...
     }
 
+    @Override
+    public String onAuthTokenRequested() {
+        IterableLogger.printInfo();
+
+        try {
+            authHandlerCallbackLatch = new CountDownLatch(1);
+            sendEvent(EventName.handleAuthTokenRequestedCalled.name(), null);
+            authHandlerCallbackLatch.await(30, TimeUnit.SECONDS);
```
Comment (vbabenkoru) [inline android/src/main/java/com/iterable/reactnative/RNIterableAPIModule.java:446]: @roninopf It looks like we don't actually call `.countDown()` on the latch anywhere, so it'll always just wait for 30 seconds.

**Review by vbabenkoru — CHANGES_REQUESTED** [review]

**Review by Ayyanchira — APPROVED** [review]
Looks good.. 🤞
Lets merge and test it out!!!

File: android/src/main/java/com/iterable/reactnative/RNIterableAPIModule.java [hunk 3]
```diff
@@ -422,9 +436,29 @@ public boolean handleIterableURL(@NonNull Uri uri, @NonNull IterableActionContex
...
+            IterableLogger.e(TAG, "auth handler module failed");
+            return null;
+        }
+    }
+
     // ---------------------------------------------------------------------------------------
     // endregion
 
+    public void passAlongAuthToken(String authToken) {
+        passedAuthToken = authToken;
```
Comment (vbabenkoru) [inline android/src/main/java/com/iterable/reactnative/RNIterableAPIModule.java:459]: Need to add something like this here
```
if (authHandlerCallbackLatch != null) {
    authHandlerCallbackLatch.countDown();
}
```

Similar to what we do in `setInAppShowResponse`.

File: ts/Iterable.ts [hunk 4]
```diff
@@ -42,39 +41,55 @@ class IterableConfig {
...
   * Implement this protocol to override default in-app behavior.
   * By default, every single in-app will be shown as soon as it is available.
   * If more than 1 in-app is available, we show the first.
   */
   inAppHandler?: (message: IterableInAppMessage) => IterableInAppShowResponse
 
+  /**
+   * The handler with which your own calls to your backend containing the auth token happen
+   */
+  authTokenRequestedHandler?: () => Promise<string | undefined>
```
Comment (vbabenkoru) [inline ts/Iterable.ts:80]: This is called `authHandler`/`authDelegate` in native SDKs. Is there a reason to have a different name here?

**Review by vbabenkoru — APPROVED** [review]
Code looks good! Haven't tested this since the dev app crashes on my emulator for some reason.

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@iterable/react-native-sdk",
-  "version": "1.0.13",
+  "version": "1.0.14-beta1",
```
Comment (vbabenkoru) [inline package.json:3]:
```suggestion
  "version": "1.0.14",
```

**Review by vbabenkoru — APPROVED** [review]
:shipit:

**Pull request merged**